### PR TITLE
Implementation of TridentNet

### PR DESCRIPTION
**Pull request by Chrisfsj2051**
Implementation of [TridentNet](https://arxiv.org/abs/1901.01892).

Please note that this implementation refers to the [implementation in detectron2](https://github.com/facebookresearch/detectron2/tree/master/projects/TridentNet), which is slightly different from the origin paper. In paper, the author use r50 to conduct abaltion study, but report final results based on r101. While in detectron2, all experiments are conducted with r50. Besides, the author use all 3 branches to yield best performance, but detectron2 only uses the middle branch (index=1, dilation=2).

This implementation is still work in progress, for slightly lower accuracy comparing to detectron2. The unit test will be added after accuracy meets expect.

| backbone | schedule  | mAP-mmdet | mAP-detr2 |
| --- | :---: | :---: | :---: |  
| r50 | 1x | 37.8 | 38.0 |
|r50 | 3x | 40.3 | 40.6 |
|r101 | 3x | Out of Memory | 43.6 |

Update 7.20: The faster-rcnn C4 baseline with 3x schedule failed to reach the accuracy in detectron2.
| type | schedule  | mAP-mmdet | mAP-detr2 |
| --- | :---: | :---: | :---: |  
| frcnn-c4-r50 | 3x | 37.8 | 38.4 |

### CODE REVIEW

**Comment by TWDH** [comment]
Hi @Chrisfsj2051 , I'm very interested in TridentNet as well. Is it runable in your repository?

**Comment by Chrisfsj2051** [comment]
Hi @TWDH , Sorry for the Late Reply. Sure, it is already runnable.

**Comment by Chrisfsj2051** [comment]
Hi @Johnson-Wang and @hellock ,

I tried to train TridentNet (as well as FRCNN-C4) with both Iteration-Based and Epoch-Based to see the difference, here's the result:

| setting  | x-based | schedule | mAP-mmdet |
| ---   | :---: | :---:    | :---:     |  
| mmdet-trident | epoch | 1x       | 37.8      |
| mmdet-trident | iter  | 1x       | 37.9      |
| detr2-trident | iter  | 1x       | **38.0**  |
| mmdet-trident | epoch | 3x       | 42.2      |
| mmdet-trident | iter  | 3x       | 42.1      |
| detr2-trident | iter  | 3x       | **42.6**      |
| mmdet-FrcnnC4 | epoch | 3x       | 38.1      |
| mmdet-FrcnnC4 | iter  | 3x       | 38.1      |
| detr2-FrcnnC4 | iter  | 3x       | **38.4**      |

As shown in this table, with iteration-based 1x schedule (~12.3 epoch), this implementation yield 37.9mAP (0.1 lower than detr2). However, under 3x schedule, TridentNet is ~0.4 lower than detr2 and Frcnn-C4 is ~0.3 lower than detr2.

I'm wondering if it is qualified to be viewed or merged?

**Comment by TWDH** [comment]
> Hi @TWDH , Sorry for the Late Reply. Sure, it is already runnable.

@Chrisfsj2051 Hi, Thanks for reply. I have a question: What's your baseline's mAP? As far as I know, the official FasterRCNN in mmedetection is already 37.8% at [https://github.com/open-mmlab/mmdetection/tree/master/configs/faster_rcnn](url). detectron2 improves the performance from 35.7 to 38. Does you code have the same improvement? Thanks, just curious.

**Comment by Chrisfsj2051** [comment]
> > Hi @TWDH , Sorry for the Late Reply. Sure, it is already runnable.
> 
> @Chrisfsj2051 Hi, Thanks for reply. I have a question: What's your baseline's mAP? As far as I know, the official FasterRCNN in mmedetection is already 37.8% at [https://github.com/open-mmlab/mmdetection/tree/master/configs/faster_rcnn](url). detectron2 improves the performance from 35.7 to 38. Does you code have the same improvement? Thanks, just curious.

Hi @TWDH , the 35.7 mAP FRCNN in detectron2 is FRCNN-C4. The 37.8 mAP FRCNN in mmdet that you mentioned above, however, is FRCNN-FPN, which has 37.9 mAP in detectron2 implementation (R50-FPN). TridentNet is based on FRCNN C4.

**Comment by hellock** [comment]
Task linked: [CU-4btyv9 TridentNet PR review & Merge](https://app.clickup.com/t/4btyv9)

**Comment by Chrisfsj2051** [comment]
> Task linked: [CU-4btyv9 TridentNet PR review & Merge](https://app.clickup.com/t/4btyv9)

It shows "You do not have access to this task code: ACCESS_083" with account "fangshijie@sensetime.com"

**Comment by cuong12ly1** [comment]
@Chrisfsj2051 I'm really interested in your work, especially ResNet101-deformable. How can i use your work? I'm a github novice. thanks

File: mmdet/models/backbones/trident_resnet.py [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+            outputs = [
+                F.conv2d(inputs[0], self.weight, self.bias, self.stride,
+                         self.paddings[self.test_branch_idx],
+                         self.dilations[self.test_branch_idx])
+            ]
+
+        return outputs
+
+
+# Since TridentNet if defined over ResNet50 and ResNet101, here we
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:92]:
```suggestion
# Since TridentNet is defined over ResNet50 and ResNet101, here we
```

File: configs/tridentnet/debug.py [hunk 2]
```diff
@@ -0,0 +1,69 @@
+_base_ = [
```
Comment (xvjiarui) [inline configs/tridentnet/debug.py:1]: This config could be removed.

File: mmdet/models/backbones/trident_resnet.py [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        self.weight = nn.Parameter(
+            torch.Tensor(out_channels, in_channels, *self.kernel_size))
+        if bias:
+            self.bias = nn.Parameter(torch.Tensor(out_channels))
+        else:
+            self.bias = None
+
+        nn.init.kaiming_uniform_(self.weight, nonlinearity='relu')
+        if self.bias is not None:
+            nn.init.constant_(self.bias, 0)
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:60]: We may use `mmcv.kaiming_init` and make this part a function `init_weight()`.

File: mmdet/models/backbones/trident_resnet.py [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+                kernel_size=1,
+                stride=conv_stride,
+                bias=False,
+            ),
+            build_norm_layer(norm_cfg, planes * block.expansion)[1]
+        ])
+        downsample = nn.Sequential(*downsample)
+
+    layers = []
+    for i in range(0, num_blocks):
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:211]: `0` is not necessary.

**Comment by Chrisfsj2051** [comment]
Hi @xvjiarui , thanks for comments!

I've reruned experiments and updated the results in configs/tridentnet/README.

File: configs/tridentnet/tridentnet_r50_caffe_1x_coco.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+    pretrained='open-mmlab://detectron2/resnet50_caffe',
+    backbone=dict(
+        type='TridentResNet',
+        trident_dilations=(1, 2, 3)
+    ),
+    roi_head=dict(
+        type='TridentRoIHead'
+    )
+)
+
```
Comment (xvjiarui) [inline configs/tridentnet/tridentnet_r50_caffe_1x_coco.py:20]:
```suggestion
        type='TridentRoIHead'))

```

File: configs/tridentnet/tridentnet_r50_caffe_1x_coco.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+        max_num=500
+    ),
+    rcnn=dict(
+        sampler=dict(
+            num=128,
+            pos_fraction=0.5,
+            add_gt_as_proposals=False
+        )
+    )
+)
```
Comment (xvjiarui) [inline configs/tridentnet/tridentnet_r50_caffe_1x_coco.py:33]:
```suggestion
            add_gt_as_proposals=False)))
```

File: configs/tridentnet/tridentnet_r50_caffe_1x_coco.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+            dict(type='ImageToTensor', keys=['img']),
+            dict(type='Collect', keys=['img']),
+        ])
+]
+data = dict(
+    train=dict(pipeline=train_pipeline),
+    val=dict(pipeline=test_pipeline),
+    test=dict(pipeline=test_pipeline))
+# optimizer
+optimizer = dict(type='SGD', lr=0.02, momentum=0.9, weight_decay=0.0001)
```
Comment (xvjiarui) [inline configs/tridentnet/tridentnet_r50_caffe_1x_coco.py:68]: This part is the same as `schedule_1x.py`?

File: mmdet/models/backbones/trident_resnet.py [hunk 8]
```diff
@@ -0,0 +1,298 @@
...
+            identity = x
+            if not isinstance(x, list):
+                x = (x, ) * num_branch
+                identity = x
+                if self.downsample is not None:
+                    identity = [self.downsample(b) for b in x]
+
+            out = [self.conv1(b) for b in x]
+            out = [self.norm1(b) for b in out]
+            out = [self.relu(b) for b in out]
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:142]: Will `multi_apply` make it easier?

File: mmdet/models/roi_heads/trident_roi_head.py [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+        det_bboxes, det_labels = [det_bboxes], [det_labels]
+
+        bbox_results = [
+            bbox2result(det_bboxes[i], det_labels[i],
+                        self.bbox_head.num_classes)
+            for i in range(len(det_bboxes))
+        ]
+
+        return bbox_results
```
Comment (xvjiarui) [inline mmdet/models/roi_heads/trident_roi_head.py:74]: If `aug_test` is not supported yet, we should raise an exception.

File: mmdet/models/detectors/trident_faster_rcnn.py [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                 roi_head,
+                 train_cfg,
+                 test_cfg,
+                 neck=None,
+                 pretrained=None):
+
+        backbone['num_branch'] = num_branch
+        backbone['test_branch_idx'] = test_branch_idx
+        roi_head['num_branch'] = num_branch
+        roi_head['test_branch_idx'] = test_branch_idx
```
Comment (xvjiarui) [inline mmdet/models/detectors/trident_faster_rcnn.py:23]: I suggest setting these in the config and assert somewhere.

File: mmdet/models/backbones/trident_resnet.py (hunk 8 above)
Comment (Chrisfsj2051) [reply mmdet/models/backbones/trident_resnet.py:139]: Using `multi_apply` doesn't seems to save space. Besides, this format is easier to understand and is composite with the following lines. So I suggest we can keep these lines unchanged.

File: mmdet/models/roi_heads/trident_roi_head.py (hunk 9 above)
Comment (Chrisfsj2051) [reply mmdet/models/roi_heads/trident_roi_head.py:72]: Now `aug_test` is supported and I'm testing the correctness.

File: configs/tridentnet/tridentnet_r50_caffe_mstrain_1x_coco.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
...
+        type='Resize',
+        img_scale=[(1333, 640), (1333, 672), (1333, 704), (1333, 736),
+                   (1333, 768), (1333, 800)],
+        multiscale_mode='value',
+        keep_ratio=True),
+    dict(type='RandomFlip', flip_ratio=0.5),
+    dict(type='Normalize', **img_norm_cfg),
+    dict(type='Pad', size_divisor=32),
+    dict(type='DefaultFormatBundle'),
+    dict(type='Collect', keys=['img', 'gt_bboxes', 'gt_labels']),
```
Comment (xvjiarui) [inline configs/tridentnet/tridentnet_r50_caffe_mstrain_1x_coco.py:19]: Additional `,`

File: mmdet/models/backbones/trident_resnet.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+
+class TridentConv(nn.Module):
+    """Trident Convolution Module.
+
+    Args:
+        in_channels (int): Number of channels in input.
+        out_channels (int): Number of channels in output.
+        kernel_size (int): Size of convolution kernel.
+        stride (int): Convolution stride. Default: 1.
+        trident_dilations (tuple[int, int, int]): Dilations of different
```
Comment (ZwwWayne) [inline mmdet/models/backbones/trident_resnet.py:20]: (tuple[int, int, int]) -> (tuple[int, int, int], optiional)

File: mmdet/models/backbones/trident_resnet.py [hunk 13]
```diff
@@ -0,0 +1,295 @@
...
+            trident branch. Default: (1, 2, 3).
+        test_branch_idx (int): In inference, all 3 branches will be used
+            if `test_branch_idx==-1`, otherwise only branch with index
+            `test_branch_idx` will be used. Default: 1.
+        bias (bool): Whether to use bias in convolution or not.
```
Comment (ZwwWayne) [inline mmdet/models/backbones/trident_resnet.py:25]: (int, optional)

File: mmdet/models/backbones/trident_resnet.py [hunk 14]
```diff
@@ -0,0 +1,295 @@
...
+
+    def __init__(
+            self,
+            in_channels,
+            out_channels,
+            kernel_size,
+            stride=1,
+            trident_dilations=(1, 2, 3),
+            test_branch_idx=1,
+            bias=False,
```
Comment (ZwwWayne) [inline mmdet/models/backbones/trident_resnet.py:37]: comma is unnecessary.

File: mmdet/models/backbones/trident_resnet.py [hunk 15]
```diff
@@ -0,0 +1,295 @@
...
+import torch.nn.functional as F
+import torch.utils.checkpoint as cp
+from mmcv.cnn import build_conv_layer, build_norm_layer, kaiming_init
+from torch.nn.modules.utils import _pair
+
+from mmdet.models.backbones.resnet import Bottleneck, ResNet
+from mmdet.models.builder import BACKBONES
+
+
+class TridentConv(nn.Module):
```
Comment (ZwwWayne) [inline mmdet/models/backbones/trident_resnet.py:12]: We may register this convolutional layer into CONV_LAYERS.

File: mmdet/models/backbones/trident_resnet.py [hunk 16]
```diff
@@ -0,0 +1,295 @@
...
+
+
+# Since TridentNet is defined over ResNet50 and ResNet101, here we
+# only support TridentBottleneckBlock.
+class TridentBottleneck(Bottleneck):
+    """BottleBlock for TridentResNet.
+
+    Args:
+        trident_dilations (tuple[int, int, int]): Dilations of different
+            trident branch. Default: (1, 2, 3).
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:99]: There is no default.

File: mmdet/models/backbones/trident_resnet.py [hunk 17]
```diff
@@ -0,0 +1,295 @@
...
+    def extra_repr(self):
+        tmpstr = 'in_channels=' + str(self.in_channels)
+        tmpstr += ', out_channels=' + str(self.out_channels)
+        tmpstr += ', kernel_size=' + str(self.kernel_size)
+        tmpstr += ', num_branch=' + str(self.num_branch)
+        tmpstr += ', test_branch_idx=' + str(self.test_branch_idx)
+        tmpstr += ', stride=' + str(self.stride)
+        tmpstr += ', paddings=' + str(self.paddings)
+        tmpstr += ', dilations=' + str(self.dilations)
+        tmpstr += ', bias=' + str(self.bias)
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:71]: We may use f-string instead.

File: mmdet/models/backbones/trident_resnet.py [hunk 18]
```diff
@@ -0,0 +1,295 @@
...
+# only support TridentBottleneckBlock.
+class TridentBottleneck(Bottleneck):
+    """BottleBlock for TridentResNet.
+
+    Args:
+        trident_dilations (tuple[int, int, int]): Dilations of different
+            trident branch. Default: (1, 2, 3).
+        test_branch_idx (int): In inference, all 3 branches will be used
+            if `test_branch_idx==-1`, otherwise only branch with index
+            `test_branch_idx` will be used. Default: 1.
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:102]: There is no default.

File: mmdet/models/backbones/trident_resnet.py [hunk 19]
```diff
@@ -0,0 +1,295 @@
...
+        concat_output (bool): Whether to concat the output list to a Tensor.
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:103]: There is no default.

File: mmdet/models/backbones/trident_resnet.py [hunk 20]
```diff
@@ -0,0 +1,295 @@
...
+                    identity = [self.downsample(b) for b in x]
+
+            out = [self.conv1(b) for b in x]
+            out = [self.norm1(b) for b in out]
+            out = [self.relu(b) for b in out]
+
+            if self.with_plugins:
+                for k in range(len(out)):
+                    out[k] = self.forward_plugin(out[k],
+                                                 self.after_conv1_plugin_names)
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:144]:
```suggestion
                    out[k] = self.forward_plugin(
                        out[k], self.after_conv1_plugin_names)
```
may looks better.

Comment (Chrisfsj2051) [reply mmdet/models/backbones/trident_resnet.py]: Agree. But yapf seems like the previous better and pre-commit hook will automatically reformat it. I temporally uninstalled it.

**Comment by Chrisfsj2051** [comment]
Using all of 3 branches for inference, we can improve the mAP of mstrain-3x setting model from 40.3 to 40.6.

File: mmdet/models/backbones/trident_resnet.py [hunk 21]
```diff
@@ -0,0 +1,292 @@
...
+                norm_cfg=norm_cfg,
+                dcn=dcn,
+                plugins=plugins,
+                test_branch_idx=test_branch_idx,
+                concat_output=True if i == num_blocks - 1 else False))
+        inplanes = planes * block.expansion
+    return nn.Sequential(*layers)
+
+
+@BACKBONES.register_module
```
Comment (xvjiarui) [inline mmdet/models/backbones/trident_resnet.py:228]:
```suggestion
@BACKBONES.register_module()
```

**Review by xvjiarui — APPROVED** [review]
